### ISSUE TITLE
empire suddenly stopped reading contents of procfile

### ISSUE THREAD

**Issue by grayaii** [remind101/empire#1036]
More to come on this as I get more info today, but we got hit by this pretty hard this.

Our version of empire client and empire controller hasn't changed in ~6 months (I'll get the version of both shortly).
As of a few days ago our emp deploys started failing.
Emp deploys started failing even with docker images that are happily running in the cluster.

The symptoms are this: the command in the procfile is being ignored.  I verified that the ECS task definition that empire is creating/updating for the service has "python2", not the shell script that is in the procfile.

My only explanation is that something recently changed in ECS that is preventing empire in updating the run command.

Our workaround is to re-build every single one of our docker images to include CMD in the Dockerfile to match that of the procfile.

More to come on this today as I get more info.

**Comment by phobologic** [comment]
A couple things:

- What's the version of emp/empire?
- Can you share the contents of the Procfile?
- Can you share the contents of the Dockerfile?
- How many apps are you working with?  Did it break for all of them?

Thanks!

**Comment by grayaii** [comment]
* Version of emp and empire: built from hash: 0575382a388cfeca9915a10bb9eb661937fe7e8b
* Contents of Procfile:
```
# cat Procfile
web:                 /code/empire-run.sh
```
* Contents of Dockerfile (I removed a ton of the RUN commands that are basically RUN mkdir, or RUN pip install, etc to remove clutter):
```
$ cat Dockerfile
FROM clearcare/jessie-base:2d3922a
RUN mkdir /root/.ssh
RUN mkdir -p /src/python
RUN mkdir /code
WORKDIR /code
RUN apt-get -y install libjpeg-dev
COPY . /code
RUN mkdir -p /code/media/tmp
```
And the DockerFile from FROM clearcare/jessie-base:2d3922a
```
FROM buildpack-deps:jessie-scm

ENV PYTHONUNBUFFERED 1

RUN apt-get update && apt-get install -y --no-install-recommends \
        binutils \
        build-essential \
        cron \
        curl \
        gcc \
        gdal-bin \
        libpq-dev \
        libxml2-dev \
        libxslt1-dev \
        locales \
        libproj-dev \
        python-dev \
        make \
        vim \
        fontconfig \
        libx11-6 \
        libxext6 \
        libxrender1 \
        rsyslog \
        xfonts-base \
        xfonts-75dpi

RUN wget -O ~/wkhtmltox.deb http://download.gna.org/wkhtmltopdf/0.12/0.12.2.1/wkhtmltox-0.12.2.1_linux-jessie-amd64.deb && \
    dpkg -i ~/wkhtmltox.deb

RUN echo "UTC" > /etc/timezone && \
    dpkg-reconfigure -f noninteractive tzdata && \
    sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && \
    echo 'LANG="en_US.UTF-8"'>/etc/default/locale && \
    dpkg-reconfigure --frontend=noninteractive locales && \
    update-locale LANG=en_US.UTF-8

ENV PYTHON_PIP_VERSION 8.1.2
RUN curl -SLk 'https://bootstrap.pypa.io/get-pip.py' | python2 \
    && pip install --no-cache-dir --upgrade pip==$PYTHON_PIP_VERSION

RUN pip install numpy==1.11.0
RUN pip install \
    django==1.6.8 \
    gevent==1.0.2 \
    lxml==2.3.4 \
    pandas==0.14.1 \
    uWSGI==2.0.3
```
* We are working with 8 apps. It broke for 2 of them.
* We use terraform to deploy empire (ie an RDS database, 2 empire controllers behind an ELB)

Some more notes:
* I competely started from scratch. I completely destroyed the empire RDS database, deleted the ECS cluster, and de-registered all 15K tasks (since there is now way to delete tasks) to start fresh.  There is now only the "default" ECS cluster.
* I then recreated the empire RDS database, created our "stage" cluster, and ran "emp deploy openresty-router"
* The openresty-router app worked!
* I then deployed the next app, that was originally causing problems, and the problem persists: It is running "/bin/bash" as the command, NOT /code/empire-run.sh:
```
[root@ip-10-20-2-175 ec2-user]# docker ps -a
CONTAINER ID        IMAGE                            COMMAND             CREATED              STATUS                          PORTS               NAMES
00b3b0f3caae        clearcare/web:1450a0d            "/bin/bash"         5 seconds ago        Exited (0) 4 seconds ago                            ecs-0ea7f12c-8a45-4ac8-a419-38433a37baa4--web-2-web-84ae92cbafe6a0810400
```
I ended up bashing onto this stopped container like so, just to snoop around (procfile is in the correct directory, along with the script that it usually runs):
```
docker commit 00b3b0f3caae foo
docker run --entrypoint=bash -it foo
# ls | grep Procfile
Procfile
# cat Procfile
web:                 /code/empire-run.sh
# pwd
/code
```
So in a nutshell, it appears the WORKDIR is correct, and the Procfile is there.
I'm just not sure why it works on one app, but no this one, and we are almost 100% certain this image  was working fine just last week.

We are looking at the extractor.go function and see what could be tripping it up.

**Comment by grayaii** [comment]
Oh, and no errors show up in the empire logs when creating the app:
```
017-01-22T20:57:55+0000 lvl=info msg=request.complete request_id=
t=2017-01-22T20:58:04+0000 lvl=info msg=PullImage request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e registry= repository=clearcare/web tag=1450a0d
t=2017-01-22T20:58:05+0000 lvl=info msg=ListServicesPages request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e
t=2017-01-22T20:58:06+0000 lvl=info msg="creating load balancer" request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e err=nil name=1383bbcc8ed0477196907019b760d9bb external=false dns-name=internal-1383bbcc8ed0477196907019b760d9bb-593636305.us-west-2.elb.amazonaws.com cert=
t=2017-01-22T20:58:08+0000 lvl=info msg=RegisterTaskDefinition request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e family=2e6d80dd-dc4a-4104-92e6-092f5965bc99--web
t=2017-01-22T20:58:08+0000 lvl=info msg=UpdateService request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e service-name=2e6d80dd-dc4a-4104-92e6-092f5965bc99--web desired-count=1 task-definition=2e6d80dd-dc4a-4104-92e6-092f5965bc99--web
t=2017-01-22T20:58:08+0000 lvl=info msg=CreateService request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e service-name=2e6d80dd-dc4a-4104-92e6-092f5965bc99--web desired-count=1 task-definition=2e6d80dd-dc4a-4104-92e6-092f5965bc99--web
t=2017-01-22T20:58:08+0000 lvl=info msg=request.complete request_id=140b25a9-ce84-4c54-9ffc-ab5a5d3dc99e
```
I then deployed the EXACT image that is working just fine in our AWS prod account to our newly created cluster and it failed!  There has to be some unknown difference that I am overlooking.

Looking at the extractor.go file, it looks like the empire controller downloads the image, creates a container, and then extracts the contents of the procFile.  I'm going to investigate that area to see if there is something funny going on there.

**Comment by grayaii** [comment]
Update: I may have found the breaking difference:
Broken:
**stage empire: Docker version 1.12.6**
stage ecs: Docker version 1.12.6
Works
**prod empire: Docker version 1.11.2**
prod ecs: Docker version 1.12.6

I'm going to investigate this.  It appears docker version 1.12.6 is breaking the empire controllers.

**Comment by phobologic** [comment]
@grayaii Curious - I'd be interested to figure out why 1.12.6 is working for some of your apps, but not others.  Did you explicitly make the change to 1.12.6?

I know that other folks (in the Slack) have had issues going to 1.12.6, and we haven't done any testing with it.  Be curious to hear what you find.

**Comment by grayaii** [comment]
@phobologic Ah ha! Everything went back to working once I downgraded the empire controllers to 1.11.2!!!  I'm guessing that the 1.12.6 version of docker packs things differently in the image/tar file and thus empire's extractor.go file does not find the procfile and falls back to using CMD.  We upgraded the version of docker in our ECS cluster a few weeks ago and we had no issues, but I did not know we upgraded the version of docker on the empire controllers, which was recently done.

I see that the version of empire that we are using is very old and the new code of empire uses a completely different way of extracting the contents of the procfile.  I'll see if we can merge that code into your local branch to fix this (I have no idea how long AWS will keep 1.11.2 in their repos, so it will good for us to fix it).

I have no idea why some apps work while others don't.  All of our Jenkins slaves that build the actual images are using 1.11.2 too. That hasn't changed.

Whew!  Once we get our empire controllers working with 1.12.6 on our end, I'll let you know what we fixed.  Good to know that other people are are feeling the pain.  We had to update to 1.12.6 in our ECS cluster because we had to update our ECS agent, and the version that we upgraded to was failing with docker 1.11.2 and thus we had to update to 1.12.6.

Feel free to close this ticket, but if we fix it, I'll definitely share the fix here!!!

**Comment by phobologic** [comment]
@grayaii Ah, ok - yeah, docker doesn't keep it's API steady between minor releases, so I'm not surprised there have been changes.  Thanks for digging into it, and glad you were able to find a work around.  Please keep us updated with what you find!
